### DISCUSSION TITLE
Multiple proposals

### DISCUSSION THREAD

**Discussion by BellezaEmporium** [iptv-org/community#769] [Ideas]
Here's a listing of proposals that could be beneficial for everyone.

1) Kodi & VLC playlist.

Kodi users have got some advantages : mpd streams with full Widevine support, certain additional things that permits more flexibility in streams...

I suggest we create a new branch called "kodi" that permits to add channels that Kodi natively supports. @Sphinxroot is a member and Kodi user, he can be more active in this area as it's his domain of expertise.

2) Group channels in m3u files by using brackets (needs additional scripting for final result)

Lots of channels have got different names for the same streams. A great idea would be to put them all in one line, followed by brackets to clearly differentiate the streams. Multilanguage streams won't be affected by this, as they're entirely different entities.

Please let me know if you have additional things to say related to these measures, or suggest changes.

**Comment by what-a** [comment]
Regarding # 2, could you provide an example of different names for the same stream?  I _think_ I know what you are describing but want to make sure.

**Reply by BellezaEmporium** [reply]
As an example, the French playlist contains 2 versions of Francophonie24

```
#EXTINF:-1 tvg-id="Francophonie24.fr" status="online",Francophonie24 (1080p)
https://5421175365ea3.streamlock.net/live/smil:switch.smil/playlist.m3u8
#EXTINF:-1 tvg-id="Francophonie.fr" status="online",Francophonie (360p)
http://mv2.tvfrancophonie.org/live/Stream1/playlist.m3u8
```

This would permit to have only this (MQA stands for Maximum Quality Available)

```
#EXTINF:-1 tvg-id="Francophonie24.fr" status="online",Francophonie24 (MQA : 1080p)
[https://5421175365ea3.streamlock.net/live/smil:switch.smil/playlist.m3u8, http://mv2.tvfrancophonie.org/live/Stream1/playlist.m3u8]
```

**Reply by Owyn** [reply]
Some players have the option to combine channels with same name into one, example "ProgTV Android"

**Reply by BellezaEmporium** [reply]
> Some players have the option to combine channels with same name into one, example "ProgTV Android"

Since the playlists are mainly remade by the bot on updates, it's just another way of seeing things up.

**Comment by what-a** [comment]
Thanks for the example of proposal # 2.  I'm guessing we might end up with some long lines when listing multiple streams on the same line. Not a deal breaker, but just want to point it out. Other than that, it seems like a good idea to me.

**Reply by BellezaEmporium** [reply]
I mean, as long as it's in a list, you can use line breaks to put each link like this 

[link1,
link2,
link3]

**Reply by what-a** [reply]
Yeah, I like that!

**Comment by freearhey** [comment]
1. With the introduction of channel feeds, a playlist can now technically contain separate links for different players. For example:

```xml
#EXTINF:-1 tvg-id="Example.us@VLC",Example TV
https://example.com/playlist.m3u8
#EXTINF:-1 tvg-id="Example.us@Kodi",Example TV
https://example.com/playlist.mpd
```

2. The links in the repository are stored in M3U playlists primarily so that they can be easily used even if the bot stops working for whatever reason and such deviation from the accepted standard will simply kill this feature.

**Reply by BellezaEmporium** [reply]
Regarding point 1, it's an excellent idea, this can bring Kodi-specific channels to our list. However, since it's got a different way of naming, we would certainly need to adapt the bot to those measures.

**Reply by freearhey** [reply]
What specific changes need to be made to the bot?

**Reply by BellezaEmporium** [reply]
I believe it would be the recognition of specific Kodi patterns.

**Reply by freearhey** [reply]
I'm sorry, I don't know much about Kodi. What patterns?

**Reply by BellezaEmporium** [reply]
No problems.

It would be 
```
#KODIPROP:
```

The most important one would certainly be 

```
#KODIPROP:inputstream=inputstream.adaptive
```

This indicates to the Kodi inputstream plugin that the stream he's going to get is an adaptive playlist. Since we made it a sort of a norm, it would need to be added to each stream.

(in this case, better not add a Kodi variant, but keep it clean and simply add the kodiprop to all known links (except chunklists and non-adaptive playlists.)

Another thing that certain players uses (and that's not linked to Kodi at all) is the "piping headers" method.

eg.
```
http(s)://abc1234.xyz/blabla.m3u8|option1=value1&option2=value2
```

Kodi natively supports it, in our example, but certain players like Tivimate and others support it as a default method of adding headers in the connection.

**Reply by CasperMcFadden95** [reply]
Hi, the Kodi pvr.iptvsimple already supports referrer/user-agent via `#EXTVLCOPT`

https://github.com/kodi-pvr/pvr.iptvsimple/blob/c9bb7f80562e20156b7c700bf4e2e465d37dc4f1/src/iptvsimple/PlaylistLoader.cpp#L198C7-L198C37

You can also force a playlist to use `inputstream.adaptive` via the `pvr.iptvsimple` settings

Do we need to make the playlist more complicated when it works?

Thanks

**Comment by ghost** [comment]
in addition to Kodi, you can also add support for Tivimate, simply by adding the referrer and other headers this way instead of using #EXTVLCOPT:

`https://link/index.m3u8|Referer="https://referer.xyz/"|User-Agent="Mozilla/5.0 (iPhone; CPU iPhone OS 17_7 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1"|Origin="https://origin.xyz" `

**Comment by freearhey** [comment]
Done in https://github.com/iptv-org/iptv/pull/24931
